contributors/design-proposals/runas-groupid.md

```diff
@@ -0,0 +1,191 @@
+    ..... 
+
+     // The UID to run the entrypoint of the container process.
+     // Defaults to user specified in image metadata if unspecified.
+     // May also be set in SecurityContext.  If set in both SecurityContext and
+     // PodSecurityContext, the value specified in SecurityContext takes precedence
+     // for that container.
+     // +optional
+     RunAsUser *int64
+     // The GID to run the entrypoint of the container process.
+     // Defaults to group specified in image metadata if unspecified.
+     // May also be set in PodSecurityContext.  If set in both SecurityContext and
+     // PodSecurityContext, the value specified in SecurityContext takes precedence.
+     // +optional
+     RunAsGroup *int64
+
+    .....
+ }
+```
+
+#### PodSecurityPolicy
+
+PodSecurityPolicy defines strategies or conditions that a pod must run with in order to be accepted
+into the system. Two of the relevant strategies are RunAsUser and SupplementalGroups. We introduce 
+a new strategy called RunAsGroup which will support the following options:
+- MustRunAs
+- MustRunAsNonRoot
+- RunAsAny
+
+```
+// PodSecurityPolicySpec defines the policy enforced.
+ type PodSecurityPolicySpec struct {
+     //Other fields not shown for brevity
+    ..... 
+  // RunAsUser is the strategy that will dictate the allowable RunAsUser values that may be set.
+  RunAsUser RunAsUserStrategyOptions
+  // SupplementalGroups is the strategy that will dictate what supplemental groups are used by the SecurityContext.
+  SupplementalGroups SupplementalGroupsStrategyOptions
+
+
+  // RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+  RunAsGroup RunAsGroupStrategyOptions
+   .....
+}
+
+// RunAsGroupStrategyOptions defines the strategy type and any options used to create the strategy.
+ type RunAsUserStrategyOptions struct {
+     // Rule is the strategy that will dictate the allowable RunAsGroup values that may be set.
+     Rule RunAsGroupStrategy
+     // Ranges are the allowed ranges of gids that may be used.
+     // +optional
+     Ranges []GroupIDRange
+ }
+
+// RunAsGroupStrategy denotes strategy types for generating RunAsGroup values for a
+ // SecurityContext.
+ type RunAsGroupStrategy string
+ 
+ const (
+     // container must run as a particular gid.
+     RunAsGroupStrategyMustRunAs RunAsGroupStrategy = "MustRunAs"
+     // container must run as a non-root gid
+     RunAsGroupStrategyMustRunAsNonRoot RunAsGroupStrategy = "MustRunAsNonRoot"
+     // container may make requests for any gid.
+     RunAsGroupStrategyRunAsAny RunAsGroupStrategy = "RunAsAny"
+ )
+```
+
+## Behavior
+
+Following points should be noted:
+
+- `FSGroup` and `SupplementalGroups` will continue to have their old meanings and would be untouched.  
+- The `RunAsGroup` In the SecurityContext will override the `RunAsGroup` in the PodSecurityContext.
+- If no `RunAsGroup` is provided in the PodSecurityContext and SecurityContext, the Group provided 
```

Comment:
I think this is desired but might be a change in behavior that we need to call out for the feature.  Right now the docker shim uses the value of `runAsUser` in `Config.User` and doesn't try to parse anything out of the image so it would ignore anything in the dockerfile.  Similar to this example when I set the user during the run command:

```
[pweil@localhost test]$ cat Dockerfile
FROM busybox
USER 0:1000
[pweil@localhost test]$ docker run pweil/group-test id
uid=0(root) gid=1000
[pweil@localhost test]$ docker run -u 1000 pweil/group-test id
uid=1000 gid=0(root)
```

The behavior above would mean the second run would have shown `1000:1000`.

Reply:
@pweil- So... we're effectively giving you group 0 whether you like it or not if you set runAsUser right now, right?

Reply:
right

Reply:
I guess just for completeness here, it's not really "we", the runtime behavior does it so there is argument there to say that you'll get whatever the runtime gives you unless we want to provide a consistent guarantee across implementations.

Also, it works the same way on the run command for the uid if only a group is set.  You're going to get root even if the Dockerfile has a UID:

```
[pweil@localhost test]$ cat Dockerfile 
FROM busybox
USER 1:1000
[pweil@localhost test]$ docker run -u :9999 pweil/group-test id
uid=0(root) gid=9999
```

Reply:
@krmayankk I think my position here would be that we do not make a guarantee to honor the GID in the dockerfile if `runAsGroup` is unset.  We guarantee to set the values you've given us and the runtime dictates the rest.  That means:

1.  `runAsUser` set to 9999, `runAsGroup` set to 9999 -> `Config.User` set to `9999:9999`
2.  `runAsUser` set to 9999, `runAsGroup` unset -> `Config.User` set to `9999` -> docker runs you with `9999:0`
3.  `runAsUser` unset, `runAsGroup` set to 9999 -> `Config.User` set to `:9999` -> docker runs you with `0:9999`

it's probably what people expect since that is current behavior from the `-u` flag and how things run now.

Reply:
@pweil- i understand the current docker behavior but is it the right behavior ?  Its a little weird to say i will only honour the Dockerfile user when -u is NOT specified , but if specified i will completely ignore the Dockerfile and take the user or group from -u flag and make the non specified one as root(0). Why not take the non specified user or group from the Dockerfile ?

Reply:
while I think that is ideal it is also a change to behavior which has backwards compatibility impacts at this point.  I don't think it's a required change to get this new api field so maybe it can be discussed more without holding up any of the progress you want to make immediately.